File: src/main/java/io/github/sunshinewzy/shining/api/namespace/Namespace.java
package io.github.sunshinewzy.shining.api.namespace;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;
import com.google.common.base.Preconditions;
import org.jetbrains.annotations.NotNull;

import java.util.Objects;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.regex.Pattern;

/**
 * The {@link #name} of namespace may only contain lowercase alphanumeric characters,
 * underscores, and hyphens.
 */
public class Namespace {

	@JsonValue
	private final String name;

	private Namespace(@NotNull String name) {
		Preconditions.checkArgument(VALID_NAMESPACE.matcher(name).matches(), "Invalid namespace. Must be [a-z0-9_-]: %s", name);
		this.name = name;
	}

	@NotNull
	public String getName() {
		return name;
	}

	@Override
	public boolean equals(Object o) {
		if (this == o) return true;
		if (o == null || getClass() != o.getClass()) return false;
		Namespace namespace = (Namespace) o;
		return Objects.equals(name, namespace.name);
	}

	@Override
	public int hashCode() {
		return name.hashCode();
	}


	public static final Pattern VALID_NAMESPACE = Pattern.compile("[a-z0-9_-]+");
	private static final ConcurrentMap<String, Namespace> cache = new ConcurrentHashMap<>();

	@NotNull
	@JsonCreator
	public static Namespace get(@NotNull String name) {
		Namespace namespace = cache.get(name);
		if (namespace != null) return namespace;

		Namespace newNamespace = new Namespace(name);
		cache.put(name, newNamespace);
		return newNamespace;
	}

}

File: src/main/kotlin/io/github/sunshinewzy/sunstcore/api/Namespace.kt
package io.github.sunshinewzy.sunstcore.api

import kotlinx.serialization.Serializable
import java.util.concurrent.ConcurrentHashMap
import java.util.regex.Pattern

/**
 * The [name] of namespace may only contain lowercase alphanumeric characters, periods,
 * underscores, and hyphens.
 */
@Serializable
class Namespace private constructor(val name: String) {

    init {
        check(VALID_NAMESPACE.matcher(name).matches()) {
            "Invalid namespace. Must be [a-z0-9._-]: $name"
        }
    }
    
    
    override fun equals(other: Any?): Boolean {
        if(this === other) return true
        if(other !is Namespace) return false

        if(name != other.name) return false

        return true
    }

    override fun hashCode(): Int {
        return name.hashCode()
    }

    override fun toString(): String {
        return name
    }


    companion object {
        private val cache = ConcurrentHashMap<String, Namespace>()
        
        val VALID_NAMESPACE = Pattern.compile("[a-z0-9._-]+")

        
        @JvmStatic
        fun get(name: String): Namespace {
            cache[name]?.let { 
                return it
            }
            
            val namespace = Namespace(name)
            cache[name] = namespace
            return namespace
        }

    }

}